src/tap/src/tapcell.cpp

```diff
@@ -33,41 +33,1306 @@
+    x_start_end.push_back(x_start);
+    x_start_end.push_back(x_end);
+    if (row_fills.find(y) == row_fills.end()) {
+      vector<vector<int>> vector2;
+      vector2.push_back(x_start_end);
+      row_fills.insert(std::pair<int, vector<vector<int>>>(y, vector2));
+    } else {
+      row_fills[y].push_back(x_start_end);
+    }
+  }
+
+  for (std::map<int, vector<vector<int>>>::iterator iter = row_fills.begin();
+       iter != row_fills.end();
+       ++iter) {
+    y = iter->first;
+    vector<int> prev_x;
+    vector<int> xs;
+    vector<vector<int>> merged_placements;
+    std::sort(row_fills[y].begin(), row_fills[y].end());
+    for (vector<int> xs : row_fills[y]) {
+      std::sort(xs.begin(), xs.end());
+      if (prev_x.size() == 0) {
+        prev_x = xs;
+      } else {
+        if (xs[0] == prev_x[1]) {
+          prev_x[1] = xs[1];
+        } else {
+          merged_placements.push_back(prev_x);
+          prev_x = xs;
+        }
+      }
+    }
+    if (prev_x.size() != 0) {
+      merged_placements.push_back(prev_x);
+    }
+    if (prev_x.size() == 0) {
+      merged_placements.push_back(xs);
+    }
+    row_fills.insert(std::pair<int, vector<vector<int>>>(y, merged_placements));
+  }
+
+  std::map<int, int> rows_with_macros;
+  std::map<std::pair<int, int>, vector<int>> macro_outlines
+      = getMacroOutlines(rows);
+  for (std::map<std::pair<int, int>, vector<int>>::iterator iter
+       = macro_outlines.begin();
+       iter != macro_outlines.end();
+       ++iter) {
+    vector<int> bot_top_row = iter->second;
+    for (int i = 0; i < bot_top_row.size(); i += 2) {
+      bot_top_row[i]--;
+      bot_top_row[i + 1]++;
+
+      if (bot_top_row[i] >= 0) {
+        rows_with_macros.insert(std::pair<int, int>(bot_top_row[i], 0));
+      }
+      if (bot_top_row[i + 1] <= rows.size()) {
+        rows_with_macros.insert(std::pair<int, int>(bot_top_row[i + 1], 0));
+      }
+    }
+  }
+
+  for (int row_idx = 0; row_idx < rows.size(); row_idx++) {
+    vector<odb::dbRow*> subrows = rows[row_idx];
+    odb::Rect rowbb;
+    subrows[0]->getBBox(rowbb);
+    const int row_y = rowbb.yMin();
+    vector<vector<int>> row_fill_check;
+    if (row_fills.find(row_y) != row_fills.end()) {
+      row_fill_check = row_fills[row_y];
+    } else {
+      row_fill_check.clear();
+    }
+
+    int gaps_above_below = 0;
+    if (rows_with_macros.begin()->first == row_idx) {
+      gaps_above_below = 1;
+      rows_with_macros.erase(rows_with_macros.begin()->first);
+    }
+
+    for (odb::dbRow* row : subrows) {
+      if (!checkSymmetry(master, row->getOrient())) {
+        continue;
+      }
+      const int site_x = row->getSite()->getWidth();
+      odb::Rect row_bb;
+      row->getBBox(row_bb);
+      int llx = row_bb.xMin();
+      const int lly = row_bb.yMin();
+      const int urx = row_bb.xMax();
+      const int ury = row_bb.yMax();
+      odb::dbOrientType ori = row->getOrient();
+      int offset = 0;
+      int pitch = -1;
```

Comment:
Please put variables as close to their first point of usage as possible (and in the smallest scope possible).